ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -531,14 +568,17 @@ void writeIndexSegmentToFile(Offset safeEndPoint) throws IOException, StoreExcep
       try {
         rwLock.readLock().lock();
 
-        // write the current version
         writer.writeShort(getVersion());
-        // write key, value size, file end pointer
-        writer.writeInt(this.keySize);
-        writer.writeInt(this.valueSize);
+        if (getVersion() == PersistentIndex.VERSION_2) {
+          writer.writeInt(getPersistedEntrySize());
+        } else {
+          // write the key size
+          writer.writeInt(getPersistedEntrySize() - getValueSize());
+        }
+        writer.writeInt(getValueSize());
         writer.writeLong(safeEndPoint.getOffset());
-        if (getVersion() == PersistentIndex.VERSION_1) {
-          // write last modified time and reset key incase of version 1
+        if (getVersion() != PersistentIndex.VERSION_0) {
+          // write last modified time and reset key in case of version 1
```

Comment:
Using `getVersion()` everywhere (except in methods that read the version off a stream or a file) as that gets overridden in the tests.